drivers/adc/ad400x/iio_ad400x.c

```diff
@@ -0,0 +1,226 @@
+* All rights reserved.
+*
+* This software is proprietary to Analog Devices, Inc. and its licensors.
+* By using this software you agree to the terms of the associated
+* Analog Devices Software License Agreement.
+*******************************************************************************/
+#ifdef IIO_SUPPORT
+
+#include <stdio.h>
+#include <math.h>
+#include <string.h>
+
+#include "ad400x.h"
+#include "iio_ad400x.h"
+#include "iio.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+
+struct ad400x_iio {
+	struct ad400x_dev *drv_dev;
+	struct iio_device *iio_dev;
+};
+
+static struct scan_type ad400x_iio_scan_types[] = {
+	[ID_AD4000] = { .sign = 'u', .realbits = 18 },
+	[ID_AD4001] = { .sign = 's', .realbits = 16 },
+	[ID_AD4002] = { .sign = 'u', .realbits = 18 },
+	[ID_AD4003] = { .sign = 's', .realbits = 18 },
+	[ID_AD4004] = { .sign = 'u', .realbits = 16 },
+	[ID_AD4005] = { .sign = 's', .realbits = 16 },
+	[ID_AD4006] = { .sign = 'u', .realbits = 18 },
+	[ID_AD4007] = { .sign = 's', .realbits = 18 },
+	[ID_AD4011] = { .sign = 's', .realbits = 18 },
+	[ID_AD4020] = { .sign = 's', .realbits = 20 },
+	[ID_ADAQ4003] = { .sign = 's', .realbits = 18},
+};
+
+static struct scan_type ad400x_iio_scan_type;
+
+/**
+ * @brief Read the debug register value
+ * @param device - Pointer to IIO device instance
+ * @param reg - Register address to read from
+ * @param val - Pointer to variable to read data into
+ * @return 0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad400x_debug_reg_read(void *device, uint32_t reg,
+		uint32_t *val)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	int ret;
+	uint8_t val8;
+
+	ret = ad400x_spi_reg_read(dev, &val8);
+	if (ret)
+		return ret;
+
+	*val = val8;
+	return 0;
+}
+
+/**
+ * @brief Write the debug register value
+ * @param device - Pointer to IIO device instance
+ * @param reg - Register address to read from
+ * @param val - Pointer to variable to write data into
+ * @return 0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad400x_debug_reg_write(void *device, uint32_t reg,
+		uint32_t val)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+
+	return ad400x_spi_reg_write(dev, (uint8_t)val);
+}
+
+static int get_raw(void *device, char *buf, uint32_t len,
+		   const struct iio_ch_info *channel, intptr_t id)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	uint32_t results = 0;
+	int32_t ret;
+
+	ret = ad400x_spi_single_conversion(dev, &results);
+	if (ret)
+		return ret;
+
+	return sprintf(buf, "%lx", results);
+}
+
+static int get_scale(void *device, char *buf, uint32_t len,
+		     const struct iio_ch_info *channel, intptr_t id)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	int32_t vals[2];
+
+	vals[0] = 5000;
```

Comment:
the reference should be given by the user (it can vary between 2.4 and 5V), so add it in iio_init_param (which has to be created and give in ad400x_iio_init function)

Reply:
yes, this is done on the last patch of this PR.. i got a separate comment that the patch that adds this should show the a user. that why its hardcoded right now but is replaced on the patch that adds ref_voltage_ms.